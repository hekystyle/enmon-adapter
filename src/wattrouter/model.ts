export enum WATTrouterModel {
  Mx = 'Mx',
}
