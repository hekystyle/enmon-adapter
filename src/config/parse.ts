import { plainToInstance } from 'class-transformer';
import { validateOrReject, ValidationError } from 'class-validator';
import { Config } from './types.js';

/**
 * @throws {ValidationError[]} if the config is invalid
 */
export const parseConfig = async (config: unknown): Promise<Config> => {
  const instance = plainToInstance(Config, config);

  try {
    await validateOrReject(instance);
  } catch (e) {
    if (Array.isArray(e) && e.every((value): value is ValidationError => value instanceof ValidationError)) {
      throw new Error(e.map(error => error.toString()).join('\n'));
    }
    throw e;
  }

  return instance;
};
