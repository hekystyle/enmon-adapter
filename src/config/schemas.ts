import { z } from 'zod';
import { configEnmonSchema, enmonIntegrationBaseConfigSchema } from '../enmon/config.schema.js';
import { WATTRouterModel } from '../wattrouter/model.enum.js';

export const THERMOMETER_MODELS = ['UNI7xxx', 'UNI1xxx'] as const;
export type ThermometerModel = (typeof THERMOMETER_MODELS)[number];

const integrationsSchema = z
  .object({
    enmon: enmonIntegrationBaseConfigSchema,
  })
  .partial();

const configThermometerSchema = z.object({
  model: z.enum(THERMOMETER_MODELS),
  dataSourceUrl: z.string().url(),
  enmon: configEnmonSchema,
});

export type ConfigThermometer = z.infer<typeof configThermometerSchema>;

const configWattrouterSchema = z.object({
  baseURL: z.string().url(),
  model: z.nativeEnum(WATTRouterModel).optional().default(WATTRouterModel.Mx),
  enmon: configEnmonSchema,
});

export type ConfigWattrouter = z.infer<typeof configWattrouterSchema>;

export const configSchema = z.object({
  integrations: integrationsSchema.optional(),
  thermometers: z.array(configThermometerSchema),
  wattrouter: configWattrouterSchema,
});

export type Config = z.infer<typeof configSchema>;
