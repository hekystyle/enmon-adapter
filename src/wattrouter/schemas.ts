import { z } from 'zod';

export const allTimeStatsSchema = z.object({
  /** Counted from this date in format YYYY-MM-DD */
  SAD: z.string(),

  /** Surplus on Phase L1 in kWh */
  SAS1: z.coerce.number(),

  /** Consumption in high tariff on Phase L1 in kWh */
  SAH1: z.coerce.number(),

  /** Consumption in low tariff on Phase L1 in kWh */
  SAL1: z.coerce.number(),

  /** Production on Phase L1 in kWh */
  SAP1: z.coerce.number(),

  /** Surplus on Phase L2 in kWh */
  SAS2: z.coerce.number(),

  /** Consumption in high tariff on Phase L2 in kWh */
  SAH2: z.coerce.number(),

  /** Consumption in low tariff on Phase L2 in kWh */
  SAL2: z.coerce.number(),

  /** Production on Phase L2 in kWh */
  SAP2: z.coerce.number(),

  /** Surplus on Phase L3 in kWh */
  SAS3: z.coerce.number(),

  /** Consumption in high tariff on Phase L3 in kWh */
  SAH3: z.coerce.number(),

  /** Consumption in low tariff on Phase L3 in kWh */
  SAL3: z.coerce.number(),

  /** Production on Phase L3 in kWh */
  SAP3: z.coerce.number(),

  /** Surplus on Phase L1+L2+L3 in kWh */
  SAS4: z.coerce.number(),

  /** Consumption in high tariff on Phase L1+L2+L3 in kWh */
  SAH4: z.coerce.number(),

  /** Consumption in low tariff on Phase L1+L2+L3 in kWh */
  SAL4: z.coerce.number(),

  /** Production on Phase L1+L2+L3 in kWh */
  SAP4: z.coerce.number(),
});

export type AllTimeStats = z.infer<typeof allTimeStatsSchema>;

export const allTimeStatsShapeSchema = z.object({
  // eslint-disable-next-line camelcase -- can't change the name of the key
  stat_alltime: allTimeStatsSchema,
});

export type AllTimeStatsShape = z.infer<typeof allTimeStatsShapeSchema>;

export const measurementSchema = z.object({
  /** Voltage on Phase L1 */
  VAC: z.coerce.number(),
});

export type MeasurementShape = z.infer<typeof measurementShapeSchema>;

export type Measurement = z.infer<typeof measurementSchema>;

export const measurementShapeSchema = z.object({
  meas: measurementSchema,
});
