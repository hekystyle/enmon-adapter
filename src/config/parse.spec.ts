import { faker } from '@faker-js/faker';
import { expect, it } from 'vitest';
import { ZodError } from 'zod';
import { type Config, type InputConfig } from './schemas.js';
import { parseConfig } from './parse.js';
import { getError } from '../../tests/utils/getError.js';
import { WATTrouterModel } from '../wattrouter/model.js';
import { EnmonEnv } from '../enmon/env.enum.js';
import { ThermometerModel } from '../thermometers/model.enum.js';

const VALID_CONFIG_INPUT_V0_1 = {
  DEV: true,
  thermometers: [
    {
      model: ThermometerModel.UNI1xxx,
      dataSourceUrl: 'http://10.0.0.0',
      enmon: {
        customerId: faker.database.mongodbObjectId(),
        devEUI: 'devEUI',
        env: EnmonEnv.Dev,
        token: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.e30.Et9HFtf9R3GEMA0IICOfFMVXY7kkTX1wr4qCyhIf58U',
      },
    },
    {
      model: ThermometerModel.UNI7xxx,
      dataSourceUrl: `http://10.0.0.0`,
      enmon: {
        customerId: faker.database.mongodbObjectId(),
        devEUI: faker.string.alphanumeric(16),
        env: EnmonEnv.Dev,
        token: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.e30.Et9HFtf9R3GEMA0IICOfFMVXY7kkTX1wr4qCyhIf58U',
      },
    },
  ],
  wattrouter: {
    baseURL: 'http://10.0.0.0',
    enmon: {
      customerId: faker.database.mongodbObjectId(),
      devEUI: 'devEUI',
      env: EnmonEnv.Dev,
      token: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.e30.Et9HFtf9R3GEMA0IICOfFMVXY7kkTX1wr4qCyhIf58U',
    },
  },
} satisfies InputConfig;

const VALID_CONFIG_OUTPUT_V0_1 = {
  DEV: VALID_CONFIG_INPUT_V0_1.DEV,
  db: { uri: 'mongodb://db/enmon-adapter' },
  enmon: {},
  thermometers: VALID_CONFIG_INPUT_V0_1.thermometers,
  wattrouters: [
    {
      model: WATTrouterModel.Mx,
      baseURL: VALID_CONFIG_INPUT_V0_1.wattrouter.baseURL,
      targets: [VALID_CONFIG_INPUT_V0_1.wattrouter.enmon],
    },
  ],
} satisfies Config;

const VALID_CONFIG_INPUT_V0_2 = {
  DEV: true,
  thermometers: [
    {
      model: ThermometerModel.UNI1xxx,
      dataSourceUrl: 'http://10.0.0.0',
      enmon: {
        customerId: faker.database.mongodbObjectId(),
        devEUI: 'devEUI',
        env: EnmonEnv.Dev,
        token: faker.internet.password(),
      },
    },
    {
      model: ThermometerModel.UNI7xxx,
      dataSourceUrl: `http://10.0.0.0`,
      enmon: {
        customerId: faker.database.mongodbObjectId(),
        devEUI: faker.string.alphanumeric(16),
        env: EnmonEnv.Dev,
        token: faker.internet.password(),
      },
    },
  ],
  wattrouter: {
    baseURL: 'http://10.0.0.0',
    enmon: {
      customerId: faker.database.mongodbObjectId(),
      devEUI: 'devEUI',
      env: EnmonEnv.Dev,
      token: faker.internet.password(),
    },
  },
} satisfies InputConfig;

const VALID_CONFIG_OUTPUT_V0_2 = {
  DEV: VALID_CONFIG_INPUT_V0_2.DEV,
  db: { uri: 'mongodb://db/enmon-adapter' },
  enmon: {},
  thermometers: VALID_CONFIG_INPUT_V0_2.thermometers,
  wattrouters: [
    {
      model: WATTrouterModel.Mx,
      baseURL: VALID_CONFIG_INPUT_V0_2.wattrouter.baseURL,
      targets: [VALID_CONFIG_INPUT_V0_2.wattrouter.enmon],
    },
  ],
} satisfies Config;

const VALID_CONFIG_INPUT_V1 = {
  DEV: true,
  db: {
    uri: faker.internet.url(),
  },
  enmon: {
    contactEmail: faker.internet.email(),
  },
  thermometers: [
    {
      model: ThermometerModel.UNI1xxx,
      dataSourceUrl: 'http://10.0.0.0',
      enmon: {
        customerId: faker.database.mongodbObjectId(),
        devEUI: 'devEUI',
        env: EnmonEnv.Dev,
        token: faker.internet.password(),
      },
    },
    {
      model: ThermometerModel.UNI7xxx,
      dataSourceUrl: `http://10.0.0.0`,
      enmon: {
        customerId: faker.database.mongodbObjectId(),
        devEUI: faker.string.alphanumeric(16),
        env: EnmonEnv.Dev,
        token: faker.internet.password(),
      },
    },
  ],
  wattrouters: [
    {
      model: WATTrouterModel.Mx,
      baseURL: 'http://10.0.0.0',
      targets: [
        {
          customerId: faker.database.mongodbObjectId(),
          devEUI: 'devEUI',
          env: EnmonEnv.Dev,
          token: faker.internet.password(),
        },
      ],
    },
  ],
} satisfies InputConfig;

const VALID_CONFIG_OUTPUT_V1 = {
  DEV: VALID_CONFIG_INPUT_V1.DEV,
  db: VALID_CONFIG_INPUT_V1.db,
  enmon: VALID_CONFIG_INPUT_V1.enmon,
  thermometers: VALID_CONFIG_INPUT_V1.thermometers,
  wattrouters: VALID_CONFIG_INPUT_V1.wattrouters,
} satisfies Config;

it.each([
  ['0.1', VALID_CONFIG_INPUT_V0_1, VALID_CONFIG_OUTPUT_V0_1],
  ['0.2', VALID_CONFIG_INPUT_V0_2, VALID_CONFIG_OUTPUT_V0_2],
  ['1', VALID_CONFIG_INPUT_V1, VALID_CONFIG_OUTPUT_V1],
])('should validate example v%s config', async (_version, input, output) => {
  const promise = parseConfig(input);

  await expect(promise).resolves.toStrictEqual(output);
});

it.each([
  [
    {
      thermometer: {},
      thermometers: [{}],
      wattrouter: {},
    },
  ],
  [
    {
      db: {},
      thermometer: { dataSourceUrl: 'invalid URL', enmon: {} },
      thermometers: [{ model: 'invalid model', dataSourceUrl: 'invalid URL', enmon: {} }],
      wattrouter: { baseURL: 'invalid URL', enmon: {} },
    },
  ],
  [
    {
      db: {
        uri: 'invalid URL',
      },
      enmon: {
        contactEmail: 'invalid email',
      },
      thermometer: {
        dataSourceUrl: 'invalid URL',
        enmon: { customerId: 'invalid ID', devEUI: '', env: 'invalid EnmonEnv', token: '' },
      },
      thermometers: [
        {
          model: 'invalid model',
          dataSourceUrl: 'invalid URL',
          enmon: { customerId: 'invalid ID', devEUI: '', env: 'invalid EnmonEnv', token: '' },
        },
      ],
      wattrouter: {
        baseURL: 'invalid URL',
        enmon: { customerId: 'invalid ID', devEUI: '', env: 'invalid EnmonEnv', token: '' },
      },
    },
  ],
])('should reject invalid config %#', async configObj => {
  const e = await getError(() => parseConfig(configObj));
  expect(e).toBeInstanceOf(ZodError);
  expect((e as ZodError).errors).toMatchSnapshot();
});
