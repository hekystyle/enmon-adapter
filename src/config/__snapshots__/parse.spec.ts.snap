// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should reject invalid config 0 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isDefined": "thermometer should not be null or undefined",
    },
    "property": "thermometer",
    "target": Config {},
    "value": undefined,
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isDefined": "wattrouter should not be null or undefined",
    },
    "property": "wattrouter",
    "target": Config {},
    "value": undefined,
  },
]
`;

exports[`should reject invalid config 1 1`] = `
[
  ValidationError {
    "children": [
      ValidationError {
        "children": [],
        "constraints": {
          "isDefined": "dataSourceUrl should not be null or undefined",
          "isString": "dataSourceUrl must be a string",
          "isUrl": "dataSourceUrl must be an URL address",
        },
        "property": "dataSourceUrl",
        "target": ConfigThermometer {},
        "value": undefined,
      },
      ValidationError {
        "children": [],
        "constraints": {
          "isDefined": "enmon should not be null or undefined",
        },
        "property": "enmon",
        "target": ConfigThermometer {},
        "value": undefined,
      },
    ],
    "property": "thermometer",
    "target": Config {
      "thermometer": ConfigThermometer {},
      "wattrouter": ConfigWattrouter {},
    },
    "value": ConfigThermometer {},
  },
  ValidationError {
    "children": [
      ValidationError {
        "children": [],
        "constraints": {
          "isDefined": "baseURL should not be null or undefined",
          "isString": "baseURL must be a string",
          "isUrl": "baseURL must be an URL address",
        },
        "property": "baseURL",
        "target": ConfigWattrouter {},
        "value": undefined,
      },
      ValidationError {
        "children": [],
        "constraints": {
          "isDefined": "enmon should not be null or undefined",
        },
        "property": "enmon",
        "target": ConfigWattrouter {},
        "value": undefined,
      },
    ],
    "property": "wattrouter",
    "target": Config {
      "thermometer": ConfigThermometer {},
      "wattrouter": ConfigWattrouter {},
    },
    "value": ConfigWattrouter {},
  },
]
`;

exports[`should reject invalid config 2 1`] = `
[
  ValidationError {
    "children": [
      ValidationError {
        "children": [],
        "constraints": {
          "isUrl": "dataSourceUrl must be an URL address",
        },
        "property": "dataSourceUrl",
        "target": ConfigThermometer {
          "dataSourceUrl": "invalid URL",
          "enmon": ConfigEnmon {},
        },
        "value": "invalid URL",
      },
      ValidationError {
        "children": [
          ValidationError {
            "children": [],
            "constraints": {
              "isDefined": "env should not be null or undefined",
              "isEnum": "env must be a valid enum value",
            },
            "property": "env",
            "target": ConfigEnmon {},
            "value": undefined,
          },
          ValidationError {
            "children": [],
            "constraints": {
              "isDefined": "customerId should not be null or undefined",
              "isMongoId": "customerId must be a mongodb id",
              "isString": "customerId must be a string",
            },
            "property": "customerId",
            "target": ConfigEnmon {},
            "value": undefined,
          },
          ValidationError {
            "children": [],
            "constraints": {
              "isDefined": "devEUI should not be null or undefined",
              "isNotEmpty": "devEUI should not be empty",
              "isString": "devEUI must be a string",
            },
            "property": "devEUI",
            "target": ConfigEnmon {},
            "value": undefined,
          },
          ValidationError {
            "children": [],
            "constraints": {
              "isDefined": "token should not be null or undefined",
              "isNotEmpty": "token should not be empty",
              "isString": "token must be a string",
            },
            "property": "token",
            "target": ConfigEnmon {},
            "value": undefined,
          },
        ],
        "property": "enmon",
        "target": ConfigThermometer {
          "dataSourceUrl": "invalid URL",
          "enmon": ConfigEnmon {},
        },
        "value": ConfigEnmon {},
      },
    ],
    "property": "thermometer",
    "target": Config {
      "thermometer": ConfigThermometer {
        "dataSourceUrl": "invalid URL",
        "enmon": ConfigEnmon {},
      },
      "wattrouter": ConfigWattrouter {
        "baseURL": "invalid URL",
        "enmon": ConfigEnmon {},
      },
    },
    "value": ConfigThermometer {
      "dataSourceUrl": "invalid URL",
      "enmon": ConfigEnmon {},
    },
  },
  ValidationError {
    "children": [
      ValidationError {
        "children": [],
        "constraints": {
          "isUrl": "baseURL must be an URL address",
        },
        "property": "baseURL",
        "target": ConfigWattrouter {
          "baseURL": "invalid URL",
          "enmon": ConfigEnmon {},
        },
        "value": "invalid URL",
      },
      ValidationError {
        "children": [
          ValidationError {
            "children": [],
            "constraints": {
              "isDefined": "env should not be null or undefined",
              "isEnum": "env must be a valid enum value",
            },
            "property": "env",
            "target": ConfigEnmon {},
            "value": undefined,
          },
          ValidationError {
            "children": [],
            "constraints": {
              "isDefined": "customerId should not be null or undefined",
              "isMongoId": "customerId must be a mongodb id",
              "isString": "customerId must be a string",
            },
            "property": "customerId",
            "target": ConfigEnmon {},
            "value": undefined,
          },
          ValidationError {
            "children": [],
            "constraints": {
              "isDefined": "devEUI should not be null or undefined",
              "isNotEmpty": "devEUI should not be empty",
              "isString": "devEUI must be a string",
            },
            "property": "devEUI",
            "target": ConfigEnmon {},
            "value": undefined,
          },
          ValidationError {
            "children": [],
            "constraints": {
              "isDefined": "token should not be null or undefined",
              "isNotEmpty": "token should not be empty",
              "isString": "token must be a string",
            },
            "property": "token",
            "target": ConfigEnmon {},
            "value": undefined,
          },
        ],
        "property": "enmon",
        "target": ConfigWattrouter {
          "baseURL": "invalid URL",
          "enmon": ConfigEnmon {},
        },
        "value": ConfigEnmon {},
      },
    ],
    "property": "wattrouter",
    "target": Config {
      "thermometer": ConfigThermometer {
        "dataSourceUrl": "invalid URL",
        "enmon": ConfigEnmon {},
      },
      "wattrouter": ConfigWattrouter {
        "baseURL": "invalid URL",
        "enmon": ConfigEnmon {},
      },
    },
    "value": ConfigWattrouter {
      "baseURL": "invalid URL",
      "enmon": ConfigEnmon {},
    },
  },
]
`;

exports[`should reject invalid config 3 1`] = `
[
  ValidationError {
    "children": [
      ValidationError {
        "children": [],
        "constraints": {
          "isUrl": "dataSourceUrl must be an URL address",
        },
        "property": "dataSourceUrl",
        "target": ConfigThermometer {
          "dataSourceUrl": "invalid URL",
          "enmon": ConfigEnmon {
            "customerId": "invalid ID",
            "devEUI": "",
            "env": "invalid EnmonEnv",
            "token": "",
          },
        },
        "value": "invalid URL",
      },
      ValidationError {
        "children": [
          ValidationError {
            "children": [],
            "constraints": {
              "isEnum": "env must be a valid enum value",
            },
            "property": "env",
            "target": ConfigEnmon {
              "customerId": "invalid ID",
              "devEUI": "",
              "env": "invalid EnmonEnv",
              "token": "",
            },
            "value": "invalid EnmonEnv",
          },
          ValidationError {
            "children": [],
            "constraints": {
              "isMongoId": "customerId must be a mongodb id",
            },
            "property": "customerId",
            "target": ConfigEnmon {
              "customerId": "invalid ID",
              "devEUI": "",
              "env": "invalid EnmonEnv",
              "token": "",
            },
            "value": "invalid ID",
          },
          ValidationError {
            "children": [],
            "constraints": {
              "isNotEmpty": "devEUI should not be empty",
            },
            "property": "devEUI",
            "target": ConfigEnmon {
              "customerId": "invalid ID",
              "devEUI": "",
              "env": "invalid EnmonEnv",
              "token": "",
            },
            "value": "",
          },
          ValidationError {
            "children": [],
            "constraints": {
              "isNotEmpty": "token should not be empty",
            },
            "property": "token",
            "target": ConfigEnmon {
              "customerId": "invalid ID",
              "devEUI": "",
              "env": "invalid EnmonEnv",
              "token": "",
            },
            "value": "",
          },
        ],
        "property": "enmon",
        "target": ConfigThermometer {
          "dataSourceUrl": "invalid URL",
          "enmon": ConfigEnmon {
            "customerId": "invalid ID",
            "devEUI": "",
            "env": "invalid EnmonEnv",
            "token": "",
          },
        },
        "value": ConfigEnmon {
          "customerId": "invalid ID",
          "devEUI": "",
          "env": "invalid EnmonEnv",
          "token": "",
        },
      },
    ],
    "property": "thermometer",
    "target": Config {
      "thermometer": ConfigThermometer {
        "dataSourceUrl": "invalid URL",
        "enmon": ConfigEnmon {
          "customerId": "invalid ID",
          "devEUI": "",
          "env": "invalid EnmonEnv",
          "token": "",
        },
      },
      "wattrouter": ConfigWattrouter {
        "baseURL": "invalid URL",
        "enmon": ConfigEnmon {
          "customerId": "invalid ID",
          "devEUI": "",
          "env": "invalid EnmonEnv",
          "token": "",
        },
      },
    },
    "value": ConfigThermometer {
      "dataSourceUrl": "invalid URL",
      "enmon": ConfigEnmon {
        "customerId": "invalid ID",
        "devEUI": "",
        "env": "invalid EnmonEnv",
        "token": "",
      },
    },
  },
  ValidationError {
    "children": [
      ValidationError {
        "children": [],
        "constraints": {
          "isUrl": "baseURL must be an URL address",
        },
        "property": "baseURL",
        "target": ConfigWattrouter {
          "baseURL": "invalid URL",
          "enmon": ConfigEnmon {
            "customerId": "invalid ID",
            "devEUI": "",
            "env": "invalid EnmonEnv",
            "token": "",
          },
        },
        "value": "invalid URL",
      },
      ValidationError {
        "children": [
          ValidationError {
            "children": [],
            "constraints": {
              "isEnum": "env must be a valid enum value",
            },
            "property": "env",
            "target": ConfigEnmon {
              "customerId": "invalid ID",
              "devEUI": "",
              "env": "invalid EnmonEnv",
              "token": "",
            },
            "value": "invalid EnmonEnv",
          },
          ValidationError {
            "children": [],
            "constraints": {
              "isMongoId": "customerId must be a mongodb id",
            },
            "property": "customerId",
            "target": ConfigEnmon {
              "customerId": "invalid ID",
              "devEUI": "",
              "env": "invalid EnmonEnv",
              "token": "",
            },
            "value": "invalid ID",
          },
          ValidationError {
            "children": [],
            "constraints": {
              "isNotEmpty": "devEUI should not be empty",
            },
            "property": "devEUI",
            "target": ConfigEnmon {
              "customerId": "invalid ID",
              "devEUI": "",
              "env": "invalid EnmonEnv",
              "token": "",
            },
            "value": "",
          },
          ValidationError {
            "children": [],
            "constraints": {
              "isNotEmpty": "token should not be empty",
            },
            "property": "token",
            "target": ConfigEnmon {
              "customerId": "invalid ID",
              "devEUI": "",
              "env": "invalid EnmonEnv",
              "token": "",
            },
            "value": "",
          },
        ],
        "property": "enmon",
        "target": ConfigWattrouter {
          "baseURL": "invalid URL",
          "enmon": ConfigEnmon {
            "customerId": "invalid ID",
            "devEUI": "",
            "env": "invalid EnmonEnv",
            "token": "",
          },
        },
        "value": ConfigEnmon {
          "customerId": "invalid ID",
          "devEUI": "",
          "env": "invalid EnmonEnv",
          "token": "",
        },
      },
    ],
    "property": "wattrouter",
    "target": Config {
      "thermometer": ConfigThermometer {
        "dataSourceUrl": "invalid URL",
        "enmon": ConfigEnmon {
          "customerId": "invalid ID",
          "devEUI": "",
          "env": "invalid EnmonEnv",
          "token": "",
        },
      },
      "wattrouter": ConfigWattrouter {
        "baseURL": "invalid URL",
        "enmon": ConfigEnmon {
          "customerId": "invalid ID",
          "devEUI": "",
          "env": "invalid EnmonEnv",
          "token": "",
        },
      },
    },
    "value": ConfigWattrouter {
      "baseURL": "invalid URL",
      "enmon": ConfigEnmon {
        "customerId": "invalid ID",
        "devEUI": "",
        "env": "invalid EnmonEnv",
        "token": "",
      },
    },
  },
]
`;
